docs/source/notes/autograd.rst

```diff
@@ -598,3 +643,83 @@ chain rule:
 
         .. math::
             \frac{\partial L}{\partial z^*} = 2 * Re(grad\_out^* * \frac{\partial s}{\partial z^{*}})
+
+.. _saved-tensors-hooks-doc:
+
+Hooks for saved tensors
+-----------------------
+
+You can control :ref:`how saved tensors are packed / unpacked
+<saved-tensors-doc>` by defining a pair of ``pack_hook`` / ``unpack_hook`` hooks.
+The ``pack_hook`` function should take a tensor as its single argument but can
+return any python object (e.g. another tensor, a tuple, or even a filename).
+The ``unpack_hook`` function takes as its single argument the output of
+``pack_hook`` and should return the original tensor to be used in the backward
+pass.
+
+An example of such pair is:
+
+.. code::
+
+    def pack_hook(x):
+        name = os.path.join(tmp_dir, str(uuid.uuid4()))
+        torch.save(x, name)
+        return name
+
+    def unpack_hook(name):
+        return torch.load(name)
+
+.. warning::
+
+    Both functions should avoid performing inplace operations on the input or
+    undefined behavior may arise.
```

Comment:
I am not a fan of have a UB here.
Why is inplace during packing forbidden?

Inplace during unpacking, we should say that this is forbidden and this function should be callable multiple times and should return the same data everytime.

Reply:
For example:

```python
def pack(t):
    t += 2
    return t - 2

torch.autograd.graph.set_saved_tensors_default_hooks(pack, lambda x: x)
a = torch.ones(5, requires_grad=True)
print(a)  # tensor([1., 1., 1., 1., 1.], requires_grad=True)
y = a * a
print(a)  # tensor([5., 5., 5., 5., 5.], requires_grad=True)
```

I guess that's a case where we can actually catch the inplace modification. The question is whether we prefer making inplace operations during pack forbidden or UB.

Reply:
It's not UB here. It actually does modify the input inplace. But it doesn't lead to any silently wrong gradients beyond that.
But I do agree that the user shouldn't do that  and we want to forbid it :D

Reply:
I'd argue that in the correct form, it's an UB because the way the backward formula was implemented can be subject to change without us notifying the user, as it is an implementation detail that they should not rely on. 
I can make a PR to always throw an error when the tensor is modified and make it very well defined behavior :)

Reply:
Here's a concrete example, powered by the fact that different backends have different derivative formulas for a given op.

```python
def f(device):
  lstm = nn.LSTM(5, hidden_size=5, num_layers=2, bidirectional=True, batch_first=True).to(device)
  for x in lstm.parameters():
    x.data.zero_()
  x = torch.zeros(5, 5, 5).to(device)
  return lstm(x)[0].sum()

print(f("cpu"))   # tensor(0., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(0., device='cuda:0', grad_fn=<SumBackward0>)

def inc(x):
  x += 1
  return x - 1

torch.autograd.graph.set_saved_tensors_default_hooks(inc, lambda x: x)
print(f("cpu"))   # tensor(2700., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(497.9154, device='cuda:0', grad_fn=<SumBackward0>)
torch.autograd.graph.reset_saved_tensors_default_hooks()
```

Collab: https://colab.research.google.com/drive/1XD9D2ack7WIAyrUYujuW0oKm8VYLkqAF?usp=sharing